src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ModuleInitializerAttributeShouldNotBeUsedInLibraries.cs

```diff
@@ -0,0 +1,80 @@
+    /// - Be accessible in the module using public or internal
+    /// </remarks>
+#pragma warning disable RS1004 // The [ModuleInitializer] feature only applies to C#
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+#pragma warning restore RS1004
+    public sealed class ModuleInitializerAttributeShouldNotBeUsedInLibraries : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2255";
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.ModuleInitializerAttributeShouldNotBeUsedInLibrariesTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.ModuleInitializerAttributeShouldNotBeUsedInLibrariesMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.ModuleInitializerAttributeShouldNotBeUsedInLibrariesDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                    s_localizableTitle,
+                                                                    s_localizableMessage,
+                                                                    DiagnosticCategory.Usage,
+                                                                    RuleLevel.BuildWarning,
+                                                                    s_localizableDescription,
+                                                                    isPortedFxCopRule: true,
+                                                                    isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                INamedTypeSymbol? moduleInitializerAttributeType = compilationContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeCompilerServicesModuleInitializerAttribute);
+
+                if (moduleInitializerAttributeType is null) return;
+
+                compilationContext.RegisterSymbolAction(context =>
+                {
+                    if (context.Symbol is IMethodSymbol method)
+                    {
+                        if (method.IsPrivate() ||
```

Comment:
Should this use `GetResultantVisibility`? The difference is when the method is declared `public`, but contained in `private` nested class.

Reply:
Good catch. According to the `ModuleInitializer` compiler feature's [accessibility tests](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Test/Symbol/Symbols/ModuleInitializers/AccessibilityTests.cs), `CS8814` will kick in on a public method within a private nested class.